### PR TITLE
[WIP] Add RFDiffusion model for protein backbone generation

### PR DESCRIPTION
**Pull request by chidwipak**
## What this PR does

This adds a diffusion model for protein backbone generation to DeepChem. I built it as a proper TorchModel subclass so it works directly with deepchem datasets.

This is my GSoC 2026 prep work for the "Implement RFDiffusion/RFDiffusion2" project.

### Model (`deepchem/models/torch_models/rfdiffusion.py`)

The main model is `RFDiffusionModel` which inherits from `TorchModel`. It implements a DDPM (denoising diffusion) for protein backbone coordinates (N, CA, C atoms).

Key parts:
- `BackboneDiffusion(nn.Module)` - the denoiser network. Transformer blocks with adaptive LayerNorm for timestep conditioning
- `CosineSchedule` - cosine variance schedule from Improved DDPM paper
- `RFDiffusionModel(TorchModel)` - wraps everything so you can just call `model.fit(dataset)` and `model.generate()`

I override `default_generator()` to handle the diffusion training logic internally (sampling random timesteps, adding noise etc). The loss is MSE between predicted noise and actual noise.

### Featurizer (`deepchem/feat/protein_backbone_featurizer.py`)

Extracts N/CA/C backbone coordinates from PDB files. Returns arrays of shape `(num_residues, 3, 3)` for each protein.

### CATH Dataset Loader (`deepchem/molnet/load_function/cath_datasets.py`)

MolNet loader that downloads PDB files from RCSB and creates train/valid/test splits. I followed the `load_pdbbind` pattern for this.

### Tests

- `test_rfdiffusion.py` - tests for model creation, training, generation, save/reload, schedule etc
- `test_protein_backbone_featurizer.py` - featurizer tests
- `test_cath_datasets.py` - loader tests

All passing.

### Examples

- `examples/rfdiffusion_demo.py` - end to end demo with CATH training and generation
- `examples/cath_dataset_usage.py` - how to use the CATH loader

## Usage

```python
import deepchem as dc

# load data
tasks, datasets, transformers = dc.molnet.load_cath(
    featurizer='ProteinBackbone', splitter='random')
train, valid, test = datasets

# train
model = dc.models.RFDiffusionModel(
    embed_dim=256, num_layers=8, num_heads=8)
model.fit(train, nb_epoch=100)

# generate new backbones
samples = model.generate(num_samples=5, seq_length=50)
```

## Training results

Tested on CATH dataset:
- Loss decreases steadily during training
- Memorization test passes (model can overfit tiny dataset)
- Generation produces valid finite coordinates

## What I want feedback on

- Is the overall approach correct for how this should fit into deepchem?
- Any issues with how I structured the TorchModel subclass?
- Should I be using the existing SE3 equivariance primitives in deepchem for the backbone?

## Next steps

- Add self-conditioning
- Try integrating deepchem existing equivariance layers
- Scale up training
- Add conditioning types (motif, length, binder)

## References

- Watson et al. "De novo design of protein structure and function with RFdiffusion." Nature 2023
- Ho et al. "Denoising diffusion probabilistic models." NeurIPS 2020
- Nichol and Dhariwal "Improved denoising diffusion probabilistic models." ICML 2021

### CODE REVIEW

**Comment by Naty89** [comment]
This is really cool work! I am also an applicant interested in extending this toward RFPeptide for cyclic peptide design. I use RFpeptide for de novo binder design so I understand this model family well. One observation i had was that the `ProteinBackboneFeaturizer` filters with `is_aa(residue, standard=True)` which would silently drop non canonical residues residues common in cyclic peptides. It also doesn't account for the N-to-C terminal covalent bond that defines cyclic topology. This could be addressed for an RFpeptide extension!
